### DISCUSSION TITLE
Multi-column Layout

### DISCUSSION THREAD

**Discussion by adaxer** [QuestPDF/QuestPDF#71] [Ideas]
Hi Marcin,

what an awesome library you created! Cudos! Hard to find any quest(!) that it cannot fulfil. One thing struck me though as I am working with it: Are you considering a multi-column layout, or is it possible to do and I just did not see it?

Usecase is as follows: Like in a newspaper the body (which in my terminology is the main content apart from all headers, footers, sidebars etc) should be rendered into two (ore more) columns, flowing from bottom of the first to top of the second, from there to the top of the next pages first column and so on. As modular the library is and with the render pass, I can imagine that it might be doable with a new element type or so...

Thanks very much, I'm curious about your thoughts
Andreas

**Comment by MarcinZiabek** [comment]
Hello, thank you for your comment 😁 

This type of layout is currently not available. The problem is, it is hard to implement in conjunction with pageable content, especially text. It may require some changes in the layouting algorithm. I am planning to take care of it in the future. Would you like to share any comments about requirements and expected functionalities?

**Reply by adaxer** [reply]
Sure. My intention would be to create sth like a product datasheet or a flyer. It could present its content with a header (containing logo, title, etc), a footer (with pagenumber, version etc.), and then a body area that shows the data with a leading paragraph and some ownermade tables with categories and key value pairs for the product attributes and specs. This area would flow in multiple columns over as many pages as needed (each of which would normally show header and footer, if not set to showonce).

So, a container like this could presumably be written like :
```
container.Stack(stack =>
{
  stack.Columns(3).Spacing(10).Separator(1).Color(Black);
});
```

This for example would put the stack items into a flow over 3 columns, with a spacing of 10 between them and a black column separator of width 1. 

I think this would be a great addition to this splendid framework, although I can imagine it being hard to layout (I did not really look much into how Skia does pdf generation, but from some superficial attempts with iText I found it quite complicated to tweak ;-))

**Reply by MarcinZiabek** [reply]
Well... what SkiaSharp does is to draw lines/rectangles/text wherever you ask it to. The whole layout magic is in the QuestPDF 😁

Let me tell you this... I am planning to add this capability but I am afraid it may require a significantly changing the layouting concept. Therefore, it is quite low on the development list and we may need to wait a couple of releases to finally get it. I hope, you understand.

**Reply by adaxer** [reply]
No problem, there are workarounds - but I'm looking forward to seeing this one day - keep up the good work :-)

**Reply by ghost** [reply]
Hello, Marcin,

thank you for this library! I find it very useful and maybe we will use it in the company where I work.

Unfortunately, I also really need the ability to create complex tables with about 22 columns. Is this possible at this time?

Thanks in advance!
@MarcinZiabek

**Reply by MarcinZiabek** [reply]
I don't expect any issues related to tables with an arbitrary number of columns. This feature is relatively scalable 😄 Please don't hesitate to share your feedback; I am always open to improving the feature set or the API.

**Reply by girlpunk** [reply]
I think the solution proposed for #729 would also work in this use case? Though that was only for text rather than all content

**Comment by DawidRobertNowak** [comment]
Hello @adaxer,

We're happy to let you know that multi-column layouts have been officially introduced in version 2024.7.0. The new `MultiColumn` element supports flowing content across columns with full paging support — exactly the newspaper-style layout you described.

You can find the documentation and usage examples here:
https://www.questpdf.com/api-reference/multi-column.html

We're closing this discussion as the feature has been delivered. Thank you for the suggestion and your patience — your feedback helped shape this feature.

Should you encounter any further problems, please don't hesitate to open a new issue.

Thank you!
